src/streaming/rules/ThroughputHistory.js

```diff
@@ -0,0 +1,186 @@
+            if (throughputDict[mediaType].length > 0 && !throughputDict[mediaType].hasCachedEntries) {
+                // already have some entries which are not cached entries
+                // prevent cached fragment loads from skewing the average values
+                return;
+            } else { // have no entries || have cached entries
+                // no uncached entries yet, rely on cached entries, set allowance for ABR rules
+                throughput /= 1000;
+                throughputDict[mediaType].hasCachedEntries = true;
+            }
+        } else if (throughputDict[mediaType] && throughputDict[mediaType].hasCachedEntries) {
+            // if we are here then we have some entries already, but they are cached, and now we have a new uncached entry
+            throughputDict[mediaType] = [];
+            latencyDict[mediaType] = [];
+        }
+
+        throughputDict[mediaType].push(throughput);
+        if (throughputDict[mediaType].length > MAX_MEASUREMENTS_TO_KEEP) {
+            throughputDict[mediaType].shift();
+        }
+
+        latencyDict[mediaType].push(latencyTimeInMilliseconds);
+        if (latencyDict[mediaType].length > MAX_MEASUREMENTS_TO_KEEP) {
+            latencyDict[mediaType].shift();
+        }
+    }
+
+    function getSampleSize(isThroughput, mediaType, isLive) {
+        let arr;
+        let sampleSize;
+
+        if (isThroughput) {
+            arr = throughputDict[mediaType];
+            sampleSize = isLive ? AVERAGE_THROUGHPUT_SAMPLE_AMOUNT_LIVE : AVERAGE_THROUGHPUT_SAMPLE_AMOUNT_VOD;
+        } else {
+            arr = latencyDict[mediaType];
+            sampleSize = AVERAGE_LATENCY_SAMPLE_AMOUNT;
+        }
+
+        if (!arr) {
+            sampleSize = 0;
+        } else if (sampleSize >= arr.length) {
+            sampleSize = arr.length;
+        } else if (isThroughput) {
+            // if throughput samples vary a lot, average over a wider sample
+            for (let i = 1; i < sampleSize; ++i) {
+                let ratio = arr[-i] / arr[-i - 1];
+                if (ratio >= THROUGHPUT_INCREASE_SCALE || ratio <= 1 / THROUGHPUT_DECREASE_SCALE) {
+                    sampleSize += 1;
+                    if (sampleSize >= arr.length) {
+                        return arr.length;
+                    }
+                }
+            }
+        }
+
+        return sampleSize;
+    }
+
+    function getAverage(isThroughput, mediaType, isDynamic) {
+        let sampleSize = getSampleSize(isThroughput, mediaType, isDynamic);
+        let dict = isThroughput ? throughputDict : latencyDict;
+        let arr = dict[mediaType];
+
+        if (sampleSize === 0 || !arr || arr.length === 0) {
+            return NaN;
+        }
+
+        arr = arr.slice(-sampleSize); // still works if sampleSize too large
+        // arr.length >= 1
+        return arr.reduce((av, elem, i) => av + (elem - av) / (i + 1));
```

Comment:
I'd prefer arr.reduce((av, elem) => av + elem) / arr.length  - more intelligible and only one divide is done.